### PR TITLE
PPO (Review only, please DO NOT merge)

### PR DESCRIPTION
**Pull request by liujch1998**

### CODE REVIEW

File: EasyLM/models/llama/llama_train_ppo.py [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+    tokenizer=LLaMAConfig.get_tokenizer_config(),
+    train_dataset=DatasetFactory.get_default_config(),
+    eval_dataset=DatasetFactory.get_default_config(),
+    optimizer=OptimizerFactory.get_default_config(),
+    checkpointer=StreamingCheckpointer.get_default_config(),
+    llama=LLaMAConfig.get_default_config(),
+    logger=mlxu.WandBLogger.get_default_config(),
+    log_all_worker=False,
+    jax_distributed=JaxDistributedConfig.get_default_config(),
+
```
Comment (hamishivi) [inline EasyLM/models/llama/llama_train_ppo.py:61]:
```suggestion
# ppo params
```

File: EasyLM/models/llama/llama_train_ppo.py [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+    _policy_model = FlaxLLaMAForCausalLMModule(
+        llama_config, dtype=get_float_dtype_by_name(FLAGS.dtype)
+    )
+    # TODO: load from pretrained ckpt
+    # policy_model = FlaxLLaMAForCausalLM.from_pretrained('meta-llama/Llama-2-7b-hf', config=llama_config, dtype=get_float_dtype_by_name(FLAGS.dtype), from_pt=True, use_auth_token=True)
+    policy_model = FlaxLLaMAForCausalLM(llama_config, dtype=get_float_dtype_by_name(FLAGS.dtype))
+    value_model = FlaxLLaMAForCausalLM(llama_config, dtype=get_float_dtype_by_name(FLAGS.dtype))
+    reference_model = FlaxLLaMAForCausalLM(llama_config, dtype=get_float_dtype_by_name(FLAGS.dtype))
+    reward_model = FlaxLLaMAForCausalLM(llama_config, dtype=get_float_dtype_by_name(FLAGS.dtype))
```
Comment (hamishivi) [inline EasyLM/models/llama/llama_train_ppo.py]: We should probably add a reward model class of some sort with a linear head, and associated training script (I'm happy to do this).

File: EasyLM/models/llama/llama_train_ppo.py [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+    generation_config = GenerationConfig(
+        max_new_tokens=32,
+        do_sample=True,
+        temperature=1.0,
+    )
+    pad_token_id = 0
+    outputs = policy_model.generate(
+        input_ids=batch['prompt_input_ids'],
+        generation_config=generation_config,
+        params=policy_train_state.params,
```
Comment (hamishivi) [inline EasyLM/models/llama/llama_train_ppo.py]: Examining the model implementation, I think doing this is perfectly fine, as long as we pass in the correct params. I'm not sure what tokens get output after the model generates an eos (but the generation still has to fill out to 32 tokens).

File: EasyLM/models/llama/llama_train_ppo.py [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+            train_state=policy_train_state,
+            gather_fns=gather_fns,
+            metadata=metadata,
+            milestone=milestone,
+        )
+        # TODO: save value model
+
+    mesh = LLaMAConfig.get_jax_mesh(FLAGS.mesh_dim)
+    with mesh:
+        policy_train_state = sharded_create_trainstate_from_params(policy_model.params)
```
Comment (hamishivi) [inline EasyLM/models/llama/llama_train_ppo.py]: My understanding is that this + the init before means we are sharded freshly-initialized models, rather than loading from a checkpoint. You should load the params using `checkpointer.load_trainstate_checkpoint` (see https://github.com/hamishivi/EasyLM/blob/main/EasyLM/models/llama/llama_train.py#L242) and then construct the train state around these (https://github.com/hamishivi/EasyLM/blob/main/EasyLM/models/llama/llama_train.py#L252). Looking at the Flax llama implementation, I think this should still work with the different wrapper, since the param names should be the same - but modifying dict keys shouldn't be too hard either.

File: EasyLM/models/llama/llama_train_ppo.py [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+    std = jnp.sqrt(jnp.sum(rewards ** 2, axis=-1, keepdims=True) / jnp.sum(mask, axis=-1, keepdims=True))
+    rewards = rewards / std
+    return rewards
+
+def reduce_mean(x, mask):
+    return jnp.sum(x * mask) / jnp.sum(mask)
+
+
+def ppo_loss(
+    policy_model, value_model,
```
Comment (hamishivi) [inline EasyLM/models/llama/llama_train_ppo.py:127]: I'm not familiar enough with PPO to be entirely sure this is correct, but seems good.

File: EasyLM/models/llama/llama_train_ppo.py [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+    batch,
+):
+    '''
+    batch: `prompt_input_ids` (B, PL), `prompt_attention_mask` (B, PL)
+    '''
+
+    # rollout from current policy, output `input_ids` (B, L), `attention_mask` (B, L), `continuation_mask` (B, L)
+    # input_ids, attention_mask, continuation_mask = batch['prompt_input_ids'], batch['prompt_attn_mask'], batch['prompt_attn_mask']
+    generation_config = GenerationConfig(
+        max_new_tokens=32,
```
Comment (hamishivi) [inline EasyLM/models/llama/llama_train_ppo.py]: 32 new tokens only? should this be a hyperparameter? Also, do the HF flax models always generate exactly 32 tokens (to avoid recompilation issues when generation lengths differ)?

File: EasyLM/models/llama/llama_train_ppo.py [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+        generation_config=generation_config,
+        params=policy_train_state.params,
+        pad_token_id=pad_token_id,
+    )
+    # print(outputs.sequences)
+    input_ids = outputs.sequences
+    attention_mask = jnp.where(input_ids == pad_token_id, 0, 1)
+    continuation_mask = jnp.concatenate([
+        jnp.zeros_like(batch['prompt_attn_mask']),
+        jnp.ones_like(attention_mask[:, batch['prompt_attn_mask'].shape[1]:]),
```
Comment (hamishivi) [inline EasyLM/models/llama/llama_train_ppo.py]: This means tokens in the continuation generated after an eos are not masked - depending on how the model is implemented this might cause issues for the rewards calculation below, since it is based on logit calcs. Can probably do something like multiply the continuation mask with the attention mask?

File: EasyLM/models/llama/llama_train_ppo.py [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+    rewards_raw = jnp.take_along_axis(rewards, last_token_index[:, None], axis=1).squeeze(-1) # (B)
+    rewards_normalized = rewards_raw * FLAGS.reward_gain + FLAGS.reward_bias # (B)
+    rewards_flattened = jnp.zeros_like(rewards)
+    rewards_flattened = rewards_flattened.at[:, last_token_index].set(rewards_normalized) # (B, L)
+    rewards_kl = continuations_logps - continuations_ref_logps # (B, L)
+    rewards_kl_penalty = FLAGS.kl_coef * rewards_kl # (B, L)
+    rewards_penalized = rewards_flattened - rewards_kl_penalty # (B, L)
+    rewards_penalized = jax.lax.stop_gradient(rewards_penalized)
+
+    for ministep in range(FLAGS.ministeps):
```
Comment (hamishivi) [inline EasyLM/models/llama/llama_train_ppo.py]: How do people usually schedule LRs for PPO with the ministep setup? In flax/optax, the step count in the train state (updated every time there is an `apply_gradients` call) is usually used to work out where the LR should be, so we might have to be careful about this. The way gradient accumulation also works is that it just updates params every time `step % accumulation_steps == 0`, which can mean the train_state steps doesn't reflect the number of gradient steps (and also means if ministeps < accumulation steps, then the model won't get updated).

**Review by hamishivi — COMMENTED** [review]
Added a bunch of comments! On the high level this looks good, my main things are around model loading and LR scheduling.

This runs fine locally for me, although I haven't thoroughly looked at the algorithmic implementation details.

File: EasyLM/models/llama/llama_train_ppo.py (hunk 2 above)
Comment (liujch1998) [reply EasyLM/models/llama/llama_train_ppo.py]: It might be possible to re-use this CausalLM class for RM, where we take the 0-th logit as the reward. After all, lm_head is a DxV linear layer, while the reward_head is a Dx1 linear layer. This shouldn't be too much overhead ...

File: EasyLM/models/llama/llama_train_ppo.py (hunk 3 above)
Comment (liujch1998) [reply EasyLM/models/llama/llama_train_ppo.py]: Cool. What I'm worried about is `policy_model` seems to keep an internal `._params` variable. Ofc this shall be overridden by the `params=` arg we pass into `generate()`. What I don't know how to do is to avoid passing in that internal variable into the jit function (when we pass in `policy_model`).

File: EasyLM/models/llama/llama_train_ppo.py (hunk 4 above)
Comment (liujch1998) [reply EasyLM/models/llama/llama_train_ppo.py]: Yes of course. Sorry I left out the ckpt loading for now. Will add this back later.
Also I suppose we can directly load from the HF Hub ckpt, e.g. `allenai/tulu-2-7b`? If so, then we probably won't need loading from local ckpt.

File: EasyLM/models/llama/llama_train_ppo.py (hunk 5 above)
Comment (liujch1998) [reply EasyLM/models/llama/llama_train_ppo.py:127]: I will double-check this part, and also ask another person familiar with PPO (maybe Ximing?) to cross-check~

File: EasyLM/models/llama/llama_train_ppo.py (hunk 6 above)
Comment (liujch1998) [reply EasyLM/models/llama/llama_train_ppo.py]: 32 is a placeholder. Will replace with a hyperparameter, and also make sure the generation is always padded to the max length.

File: EasyLM/models/llama/llama_train_ppo.py (hunk 8 above)
Comment (liujch1998) [reply EasyLM/models/llama/llama_train_ppo.py]: I think for PPO we should get rid of the `accumulation_steps` and update every step. As for LR, AFAIK people tend to use a constant LR, so we don't need to worry too much about tracking exact steps.

File: EasyLM/models/llama/llama_train_ppo.py (hunk 2 above)
Comment (hamishivi) [reply EasyLM/models/llama/llama_train_ppo.py]: Yeah, agree there. I think that having a reward class that wraps this behaviour is still a good idea, though, since then it means adding other reward model implementations (e.g., you generate a score or ping gpt-4 or something weird) is easier.

File: EasyLM/models/llama/llama_train_ppo.py (hunk 4 above)
Comment (hamishivi) [reply EasyLM/models/llama/llama_train_ppo.py]: Can't load directly from HF I think, but you can find easylm compatible checkpoints on huggingface (https://huggingface.co/openlm-research/open_llama_7b_easylm), or alternatively I have some on a google bucket (that you can just point to directly and the code will auto-download for you).

File: EasyLM/models/llama/llama_train_ppo.py (hunk 8 above)
Comment (hamishivi) [reply EasyLM/models/llama/llama_train_ppo.py]: Easy, thats good! Then we need to set some stuff in the optimizer to make it a constant LR - init_lr and end_lr should be the same
